### PR TITLE
[Remote Segment Store] Use files from segmentInfosSnapshot in upload

### PR DESCRIPTION
**Pull request by sachinpkale**
### Description
- `RemoteStoreRefreshListener` is one of the refresh listeners that uploads segments to remote segment store.
- Currently, it reads `segmentInfosSnapshot` from InternalEngine and adds files referenced by `segments_N` file.
- This is a defensive approach that was taken so that we do not lose track of any segment file but this can create issue in scenarios such as forcemerge, where segmentInfosSnapshot is referencing to segments_N file but the files referenced by segments_N are merged away.
- In this PR, we use segment files referenced by segmentInfosSnapshot alone.
- Added integ tests around the forcemerge use case. Existing integ tests should pass.

### Check List
- [x] New functionality includes testing.
  - [x] All tests pass
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [x] Commits are signed per the DCO using --signoff
- [ ] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by sachinpkale** [comment]
Taking a look at test failures in `org.opensearch.index.shard.RemoteStoreRefreshListenerTests`

File: server/src/internalClusterTest/java/org/opensearch/remotestore/RemoteStoreForcemergeIT.java [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+    public void setup() {
+        setupRepo();
+    }
+
+    @Override
+    public Settings indexSettings() {
+        return remoteStoreIndexSettings(0);
+    }
+
+    private IndexResponse indexSingleDoc() {
```
Comment (mch2) [inline server/src/internalClusterTest/java/org/opensearch/remotestore/RemoteStoreForcemergeIT.java:55]: This same method is in `RemoteStoreIT` and `RemoteStoreStatsIT`, lets move to `RemoteStoreBaseIntegTestCase` ?

File: server/src/internalClusterTest/java/org/opensearch/remotestore/RemoteStoreForcemergeIT.java [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+        client().admin().cluster().restoreRemoteStore(new RestoreRemoteStoreRequest().indices(INDEX_NAME), PlainActionFuture.newFuture());
+        ensureGreen(INDEX_NAME);
+
+        verifyRestoredData(indexStats, deleteDocs);
+    }
+
+    // Following integ tests use randomBoolean to control the number of integ tests. If we use the separate
+    // values for each of the flags, number of integ tests become 16 in comparison to current 2.
+    // We have run all the 16 tests on local and they run fine.
+    public void testRestoreForcemegeSingleIteration() throws IOException {
```
Comment (mch2) [inline server/src/internalClusterTest/java/org/opensearch/remotestore/RemoteStoreForcemergeIT.java:130]:
```suggestion
    public void testRestoreForceMergeSingleIteration() throws IOException {
```

File: server/src/internalClusterTest/java/org/opensearch/remotestore/RemoteStoreForcemergeIT.java [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+        boolean invokeFLush = randomBoolean();
+        boolean flushAfterMerge = randomBoolean();
+        boolean deleteDocs = randomBoolean();
+        testRestoreWithMergeFlow(1, invokeFLush, flushAfterMerge, deleteDocs);
+    }
+
+    public void testRestoreForcemegeMultipleIterations() throws IOException {
```
Comment (mch2) [inline server/src/internalClusterTest/java/org/opensearch/remotestore/RemoteStoreForcemergeIT.java:137]:
```suggestion
    public void testRestoreForceMergeMultipleIterations() throws IOException {
```

File: server/src/internalClusterTest/java/org/opensearch/remotestore/RemoteStoreForcemergeIT.java [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import static org.opensearch.test.hamcrest.OpenSearchAssertions.assertAcked;
+import static org.opensearch.test.hamcrest.OpenSearchAssertions.assertHitCount;
+
+@OpenSearchIntegTestCase.ClusterScope(scope = OpenSearchIntegTestCase.Scope.TEST, numDataNodes = 0)
```
Comment (mch2) [inline server/src/internalClusterTest/java/org/opensearch/remotestore/RemoteStoreForcemergeIT.java:33]: Can we use suite scope here instead with numDataNodes = 3?  Will speed up test execution and provide assertions that all tasks are closed post test execution.

**Comment by sachinpkale** [comment]
New integ tests are executed 180+ times on local to avoid checking in flaky tests.

**Comment by sachinpkale** [comment]
`org.opensearch.index.shard.RemoteStoreRefreshListenerTests*` should not fail.

File: server/src/main/java/org/opensearch/index/shard/RemoteStoreRefreshListener.java [hunk 5]
```diff
@@ -218,81 +214,56 @@ private synchronized void syncSegments(boolean isRetry) {
...
-                                    // integration.
-                                    logger.warn("Exception in post new segment upload actions", e);
-                                } finally {
-                                    doComplete(shouldRetry);
-                                }
+                    // Create a map of file name to size and update the refresh segment tracker
+                    updateLocalSizeMapAndTracker(localSegmentsPostRefresh);
+
+                    // Create a map of file name to size and update the refresh segment tracker
+                    updateLocalSizeMapAndTracker(localSegmentsPostRefresh);
```
Comment (mch2) [inline server/src/main/java/org/opensearch/index/shard/RemoteStoreRefreshListener.java:221]: Looks like an unintentional duplicate call here?

File: server/src/main/java/org/opensearch/index/shard/RemoteStoreRefreshListener.java [hunk 6]
```diff
@@ -218,81 +214,56 @@ private synchronized void syncSegments(boolean isRetry) {
...
+                                    checkpoint
+                                );
+                                // At this point since we have uploaded new segments, segment infos and segment metadata file,
+                                // along with marking minSeqNoToKeep, upload has succeeded completely.
+                                shouldRetry = false;
+                            } catch (Exception e) {
+                                // We don't want to fail refresh if upload of new segments fails. The missed segments will be re-tried
+                                // in the next refresh. This should not affect durability of the indexed data after remote trans-log
+                                // integration.
+                                logger.warn("Exception in post new segment upload actions", e);
```
Comment (mch2) [inline server/src/main/java/org/opensearch/index/shard/RemoteStoreRefreshListener.java:245]: nit on this comment - Is this retried on the next refresh or immediately?  This *will* impact the computation of staleness for replicas for segrep backpressure so we don't to wait until a future refresh.

**Comment by mch2** [comment]
Found an occurance of this while trying to debug flaky test testDropPrimaryDuringReplication.  NoSuchFile hit on the primary during upload.

```

java.nio.file.NoSuchFileException: /private/var/folders/c0/63hvxjzj32j_zjnrs_b5kqg80000gs/T/org.opensearch.remotestore.SegmentReplicationUsingRemoteStoreIT_4220E6BD90A7660A-013/tempDir-002/node_t5/nodes/0/indices/lUTbvNOZR0e4Y7rUmE7xuQ/0/index/segments_4
	at sun.nio.fs.UnixException.translateToIOException(UnixException.java:92) ~[?:?]
	at sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:106) ~[?:?]
	at sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:111) ~[?:?]
	at sun.nio.fs.UnixFileSystemProvider.newFileChannel(UnixFileSystemProvider.java:181) ~[?:?]
	at org.apache.lucene.tests.mockfile.FilterFileSystemProvider.newFileChannel(FilterFileSystemProvider.java:204) ~[lucene-test-framework-9.7.0.jar:9.7.0 ccf4b198ec328095d45d2746189dc8ca633e8bcf - 2023-06-21 11:48:16]
	at org.apache.lucene.tests.mockfile.DisableFsyncFS.newFileChannel(DisableFsyncFS.java:44) ~[lucene-test-framework-9.7.0.jar:9.7.0 ccf4b198ec328095d45d2746189dc8ca633e8bcf - 2023-06-21 11:48:16]
	at org.apache.lucene.tests.mockfile.FilterFileSystemProvider.newFileChannel(FilterFileSystemProvider.java:204) ~[lucene-test-framework-9.7.0.jar:9.7.0 ccf4b198ec328095d45d2746189dc8ca633e8bcf - 2023-06-21 11:48:16]
	at org.apache.lucene.tests.mockfile.HandleTrackingFS.newFileChannel(HandleTrackingFS.java:171) ~[lucene-test-framework-9.7.0.jar:9.7.0 ccf4b198ec328095d45d2746189dc8ca633e8bcf - 2023-06-21 11:48:16]
	at org.apache.lucene.tests.mockfile.HandleTrackingFS.newFileChannel(HandleTrackingFS.java:171) ~[lucene-test-framework-9.7.0.jar:9.7.0 ccf4b198ec328095d45d2746189dc8ca633e8bcf - 2023-06-21 11:48:16]
	at java.nio.channels.FileChannel.open(FileChannel.java:304) ~[?:?]
	at java.nio.channels.FileChannel.open(FileChannel.java:363) ~[?:?]
	at org.apache.lucene.store.NIOFSDirectory.openInput(NIOFSDirectory.java:78) ~[lucene-core-9.7.0.jar:9.7.0 ccf4b198ec328095d45d2746189dc8ca633e8bcf - 2023-06-21 11:48:16]
	at org.opensearch.index.store.FsDirectoryFactory$HybridDirectory.openInput(FsDirectoryFactory.java:166) ~[classes/:?]
	at org.apache.lucene.store.FilterDirectory.openInput(FilterDirectory.java:101) ~[lucene-core-9.7.0.jar:9.7.0 ccf4b198ec328095d45d2746189dc8ca633e8bcf - 2023-06-21 11:48:16]
	at org.apache.lucene.store.FilterDirectory.openInput(FilterDirectory.java:101) ~[lucene-core-9.7.0.jar:9.7.0 ccf4b198ec328095d45d2746189dc8ca633e8bcf - 2023-06-21 11:48:16]
	at org.apache.lucene.store.Directory.openChecksumInput(Directory.java:156) ~[lucene-core-9.7.0.jar:9.7.0 ccf4b198ec328095d45d2746189dc8ca633e8bcf - 2023-06-21 11:48:16]
	at org.apache.lucene.index.SegmentInfos.readCommit(SegmentInfos.java:298) ~[lucene-core-9.7.0.jar:9.7.0 ccf4b198ec328095d45d2746189dc8ca633e8bcf - 2023-06-21 11:48:16]
	at org.apache.lucene.index.SegmentInfos.readCommit(SegmentInfos.java:289) ~[lucene-core-9.7.0.jar:9.7.0 ccf4b198ec328095d45d2746189dc8ca633e8bcf - 2023-06-21 11:48:16]
	at org.opensearch.index.shard.RemoteStoreRefreshListener.syncSegments(RemoteStoreRefreshListener.java:236) ~[classes/:?]
	at org.opensearch.index.shard.RemoteStoreRefreshListener.lambda$afterRefresh$0(RemoteStoreRefreshListener.java:183) ~[classes/:?]
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:577) [?:?]
	at java.util.concurrent.FutureTask.run(FutureTask.java:317) [?:?]
	at org.opensearch.common.util.concurrent.ThreadContext$ContextPreservingRunnable.run(ThreadContext.java:849) [classes/:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144) [?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642) [?:?]
	at java.lang.Thread.run(Thread.java:1589) [?:?]
[2023-07-14T08:29:00,584][ERROR][o.o.i.s.RemoteStoreRefreshListener] [node_t5] [test-idx-1][0] segmentInfosFiles=[_1.cfs, _0.cfe, _0.si, _1.cfe, _1.si, _0.cfs, segments_4] diskFiles=[_0.cfe, _0.cfs, _0.si, _1.cfe, _1.cfs, _1.si, segments_5, write.lock]
[2023-07-14T08:29:00,584][WARN ][o.o.i.s.RemoteStoreRefreshListener] [node_t5] [test-idx-1][0] Exception while uploading new segments to the remote segment store
```

File: server/src/main/java/org/opensearch/index/shard/RemoteStoreRefreshListener.java (hunk 5 above)
Comment (sachinpkale) [reply server/src/main/java/org/opensearch/index/shard/RemoteStoreRefreshListener.java:221]: Yes, would have got added as part of resolving merge conflicts. Thanks for catching this.

File: server/src/main/java/org/opensearch/index/shard/RemoteStoreRefreshListener.java (hunk 6 above)
Comment (sachinpkale) [reply server/src/main/java/org/opensearch/index/shard/RemoteStoreRefreshListener.java:245]: It is retried with exponential back-off strategy which starts with 1s interval.

**Comment by sachinpkale** [comment]
Fixing merge conflict with main

**Comment by sachinpkale** [comment]
`org.opensearch.indices.replication.RemoteStoreReplicationSourceTests.testGetCheckpointMetadataEmpty` test is failing, looking into it.

**Comment by sachinpkale** [comment]
> `org.opensearch.indices.replication.RemoteStoreReplicationSourceTests.testGetCheckpointMetadataEmpty` test is failing, looking into it.

Re-basing `main` fixed the problem.

File: server/src/internalClusterTest/java/org/opensearch/remotestore/RemoteStoreForceMergeIT.java [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+            totalOperations += numberOfOperations;
+            if (invokeFlush) {
+                flush(INDEX_NAME);
+            } else {
+                refresh(INDEX_NAME);
+            }
+        }
+        if (deleteDocs) {
+            for (int j = 0; j < 10; j++) {
+                int length = indexResponseList.size();
```
Comment (ashking94) [inline server/src/internalClusterTest/java/org/opensearch/remotestore/RemoteStoreForceMergeIT.java:74]: nit - lets move this outside the for.

**Review by ashking94 — APPROVED** [review]
LGTM.

File: server/src/internalClusterTest/java/org/opensearch/remotestore/RemoteStoreForceMergeIT.java [hunk 8]
```diff
@@ -0,0 +1,134 @@
...
+    // values for each of the flags, number of integ tests become 16 in comparison to current 2.
+    // We have run all the 16 tests on local and they run fine.
+    public void testRestoreForceMergeSingleIteration() throws IOException {
+        boolean invokeFLush = randomBoolean();
+        boolean flushAfterMerge = randomBoolean();
+        boolean deleteDocs = randomBoolean();
+        testRestoreWithMergeFlow(1, invokeFLush, flushAfterMerge, deleteDocs);
+    }
+
+    public void testRestoreForceMergeMultipleIterations() throws IOException {
```
Comment (ashking94) [inline server/src/internalClusterTest/java/org/opensearch/remotestore/RemoteStoreForceMergeIT.java:128]: Let's add a case where all the documents are deleted.

Comment (ashking94) [inline server/src/internalClusterTest/java/org/opensearch/remotestore/RemoteStoreForceMergeIT.java:128]: Let's add a case where all the documents are deleted.

File: server/src/internalClusterTest/java/org/opensearch/remotestore/RemoteStoreForceMergeIT.java [hunk 9]
```diff
@@ -0,0 +1,139 @@
...
+    }
+
+    @Before
+    public void setup() {
+        setupRepo();
+    }
+
+    @Override
+    public Settings indexSettings() {
+        return remoteStoreIndexSettings(0);
```
Comment (mch2) [inline server/src/internalClusterTest/java/org/opensearch/remotestore/RemoteStoreForceMergeIT.java:51]: Should we have versions of these with replicas?

File: server/src/internalClusterTest/java/org/opensearch/remotestore/RemoteStoreForceMergeIT.java [hunk 10]
```diff
@@ -0,0 +1,139 @@
...
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import static org.opensearch.test.hamcrest.OpenSearchAssertions.assertAcked;
+import static org.opensearch.test.hamcrest.OpenSearchAssertions.assertHitCount;
+
+@OpenSearchIntegTestCase.ClusterScope(scope = OpenSearchIntegTestCase.Scope.TEST, numDataNodes = 3)
```
Comment (mch2) [inline server/src/internalClusterTest/java/org/opensearch/remotestore/RemoteStoreForceMergeIT.java:32]: nit - use SUITE scope instead to speed this up?

**Review by mch2 — APPROVED** [review]

File: server/src/internalClusterTest/java/org/opensearch/remotestore/RemoteStoreForceMergeIT.java (hunk 9 above)
Comment (sachinpkale) [reply server/src/internalClusterTest/java/org/opensearch/remotestore/RemoteStoreForceMergeIT.java:51]: Currently, these tests focus more on restore flow. With replicas, it becomes fail-over flow which should be covered in the extension of SegRep tests with remote store settings.

Comment (sachinpkale) [reply server/src/internalClusterTest/java/org/opensearch/remotestore/RemoteStoreForceMergeIT.java:51]: Let me check and get back on this. But mostly will create another PR to add tests with replicas.

**Pull request merged**